GPU Cholesky decomposition op

matt-graham opened the pull request:
## Reason for PR

There is currently no GPU Cholesky decomposition op to correspond to the `scipy.linalg` based `slingalg.Cholesky` op, and no existing PR implementing one as far as I can see (though an implementation of one might be useful for #2059). 

Cholesky decompositions are among other things useful in allowing more numerically stable and efficient solving of positive definite matrix systems (particularly if solving multiple systems sequentially as it allows an initial expensive decomposition to be reused without the normal numerical issues involved in precomputing a direct matrix inverse) and for use in generating samples from a multivariate Gaussian with a specified covariance matrix. Both of these uses crop up quite a lot in machine learning and statistics problems, and the O(D<sup>3</sup>) cost of computing the Cholesky decomposition means that such operations can often be a dominant cost in an algorithm so having a GPU accelerated operation can potentially give big gains.

## Description of proposed changes

This PR adds a CULA based GPU Cholesky decomposition op `GpuCholesky` which complements the exisiting CULA based `GpuSolve` op. 

For consistency with the existing `slinalg.Cholesky` op this returns a matrix in which only the entries below or above the diagonal (depending on value of `lower` parameter) are non-zero. This involves some further calls to `pycuda` based functions in `skcuda.linalg` for setting the above / below diagonal elements to zero as the CULA Cholesky decomposition function `culaDeviceSpotrf` acts destructively on the matrix it is decomposing and assigns the decomposition to the lower or upper triangular elements while leaving the existing values in the other elements in place (this is also the case for other LAPACK `*potrf` based Cholesky implementations e.g. `scipy.linalg.cho_factor`).

As well as the `GpuCholesky` op a corresponding `local_gpu_cholesky` optimization has been added which replaces `slinalg.Cholesky` op instances with `GpuCholesky` were possible. Unit tests for both the new op and optimization have also been added.

## Questions

  * The op currently creates a copy of the input matrix before performing the decomposition due to the destructive behaviour of the CULA `spotrf` function described above. In some cases maybe the original matrix does not need to be retained so an inplace / destructive operation would be permissible.   Would it be worth implementing this? Currently the corresponding `slinalg.Cholesky` op has a flag for this but it seems to not be implemented. If it was worthwhile adding this, are there any other requirements than defining a `destroy_map` attribute for the op as described [here](http://deeplearning.net/software/theano/extending/inplace.html) to make this work correctly (as well as obviously implementing the logic to do the operation in place)?

nouiz commented:
The inplace version would be useful to remove the copy. So it would speed it up a bit. how much depend of the rest of the code.

This need the parameter and small code change, but this also need an optimization that would convert the non-inplace version of the op to the inplace version of the op.

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -710,6 +710,34 @@ def local_gpu_solve(node):
+
+    if isinstance(node.op, slinalg.Cholesky):
+        if any([i.owner and isinstance(i.owner.op, HostFromGpu)
+                for i in node.inputs]):
```
As the op always have only 1 input, you can remove the list comprehension and do the check only on node.inputs[0]. This make the code more clear.

matt-graham replied:
Cheers will update, bad case of copying and pasting code from another op without really considering what it was doing!

nouiz commented on theano/sandbox/cuda/cula.py:
```diff
@@ -140,3 +143,142 @@ def cula_gpu_solve(A_, b_, trans='T'):
+        inp = as_cuda_ndarray_variable(inp)
+
+        assert inp.ndim == 2
+        return theano.Apply(self, [inp], [self.output_type(inp)()])
```
I would remove the output_type method and use use inp.type() for the output. So replace [self.output...] by [inp.type()]

matt-graham replied:
Thanks, will change, I hadn't fully understood what this code was doing.

nouiz commented on theano/sandbox/cuda/cula.py:
```diff
@@ -140,3 +143,142 @@ def cula_gpu_solve(A_, b_, trans='T'):
+            # decomposition to lower / upper triangle of input array therefore
+            # force copy of array with allocation of a new C-contiguous buffer.
+            # CULA operation assumes Fortran ordering however input matrix must
+            # be symmetric therefore no need to alter order.
```
Why inputs must be symmetric? I don't think this is right. I could easily add a parameter order that default to C, but could be F as numpy. This would make the copy a fortran order version.

matt-graham replied:
The Choleksy decomposition is only defined for positive definite matrices, which for real matrices are a subset of the symmetric matrices. If a non-symmetric and so non-positive-definite input is passed the underlying CULA function will raise an exception as the Cholesky decomposition algorithm will fail. Given an exception will be raised anyway without any additional checking overhead for non-symmetric inputs it didn't seem worth adding any explicit checks for symmetry. 

I guess the comment is not strictly correct however so I'll edit to make it more clear - perhaps it would be better to say any valid input will be symmetric? It is possible a non-symmetric matrix input will be passed to the op however it does not matter whether it uses C or Fortan ordering here as in either case the matrix will remain non-symmetric and end up raising an exception when it is attempted to calculate the decomposition. 

I'll also add an unit test to ensure non-symmetric inputs cause an exception to be raised in addition to the current one for a non-positive-definite but symmetric input.

nouiz commented:
Other then my comments, all looks pretty good. thanks.

matt-graham commented:
With regards to the inplace version:

  * The `skcuda.cula` `culaDeviceSpotrf` and the `skcuda.linalg` `triu` and `tril` functions both require / assume contiguous input arrays I believe. If I use `basic_ops.gpu_contiguous` to ensure a contiguous copy is made if necessary when `inplace = True` will this cause some sort of inconsistency as the op in no longer actually in place? Or alternatively is there anyway to force no inplace op to be used in these settings? 

  * As there is also a `lower` property for the op, (how) should I define op instances? For example should I just add one for each of the possible property values e.g `gpu_cholesky_lower_no_inplace` etc. - it seems this would get a bit unmanageable for a large number of properties however I assume there is some overhead to creating instances on the fly using e.g. `GpuCholesky(lower=True, inplace=False)`? And would it make sense therefore to define an in place optimisation for each of the different `lower` values? For now I have defined instances for each `lower` / `inplace` combination and a separate inplace optimisation for each `lower` value.

  * Is there a standard way to define tests on inplace ops? My attempts caused a series of `TypeError("Graph must not contain inplace operations")` exceptions to be raised so I have removed them for now.

abergeron commented:
If you wrap your input in gpu_contiguous, it will not cause problems for inplace since that is just another op that will execute before yours.

You don't need to define op instances.  Even if lot of the ops do, it's just for quick access.  If it doesn't make sense for yours, then don't do it.  There is a minimal overhead to creating new instances, but it should be trivial and is also not paid during execution.

You can pass accept_inplace=True to function() for testing purposes.  This is not documented because we don't want people to use it in general as you may provide unsafe graphs with this.

A better way would be to generate some small graph where the operation should be inplace and verify in the compiled graph that it is inplace.

matt-graham commented:
Now updated to remove op instance definitions and add in tests of inplace op and the associated optimization.

I defined a small graph to verify the inplace operation is applied successfully as suggested (in `tests/test_opt.py`) but also used `accept_inplace=True` to check the inplace operations output was consistent with Numpy in `tests/test_cula.py` as it seemed cleaner to test the validity of the operation implementation and optimization separately.

I wasn't sure quite how to handle registering the inplace optimization  - I added a new `optdb.register` call for CULA inplace ops rather than wrapping this in to the existing call registering the inplace BLAS ops. In particular I wasn't sure how to choose the priority so just used the same value (70.0) as for the BLAS ops.

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -710,6 +711,35 @@ def local_gpu_solve(node):
+            isinstance(host_input.owner.op,
+                       slinalg.Cholesky)):
+            A = host_input.owner.inputs[0]
+            A_ndarray = as_cuda_ndarray_variable(A)
```
ndarray make me think it is on the CPU. What about using A_gpu?

nouiz commented on theano/sandbox/cuda/opt.py:
```diff
@@ -2032,6 +2062,16 @@ def local_inplace_ger(node):
+@local_optimizer([GpuCholesky(lower=True, inplace=False),
+                  GpuCholesky(lower=False, inplace=False)], inplace=True)
+def local_inplace_gpu_cholesky(node):
+    if node.op == GpuCholesky(lower=True, inplace=False):
```
The full code of this method can be simplified:

if isinstance(node.op, GpuCholesky) and not node.op.inplace:
    return [GpuCholesky(node.op.lower, inplace=True)]

Constructing object just to compare is time consuming and would make optimization slower. So this code will be simpler and faster.

nouiz commented on theano/sandbox/cuda/tests/test_cula.py:
```diff
@@ -67,3 +67,111 @@ def test_uni_rand_solve(self):
+            A_val = numpy.random.normal(size=(3, 2)).astype("float32")
+            fn = self.get_gpu_cholesky_func(True, False)
+            fn(A_val)
+        self.assertRaises(ValueError, invalid_input_func)
```
This can be simplified. You don't need to make a new method. So just do something like:

def test_...(self):
    A_val = ...
   fn = ...
   self.assertRaises(ValueError, fn, A_val).

idem for the other tests.

nouiz commented:
thanks PR seem to advance well. I just got a few comments. I killed some of travis run, as it would have taken it too much time to test all of them.

thanks.

matt-graham commented:
Some further updates in response to comments.

@nouiz : I wasn't sure what your comment

> You can just add GpuCholesky. This will work for all instance.

 was referring to? Is this in reference to the `tracks` argument being passed to the `local_optimizer` decorator which currently adds two separate `GpuCholesky` instances?

Apologies about the large amounts of commits and so Travis runs; I am using a remote GPU server to run the tests as I don't have any CUDA compatible devices on my development machine, so I've being using git to sync the code to the remote server. Is there a simple way for me to switch to manually updating the PR after I make a series of pushes to my fork branch rather than it tracking every new push? Sorry if this is a stupid question, my Google searches didn't turn up an obvious answer.

lamblin commented:
Regarding the PR updates, the simplest way I can think of would be for you to have two remote branches on GitHub: one being tracked by the PR, and one for synchronization between your machines, say `gpu_cholesky_sync`.
You can still have only one branch on your machines, but decide to push to one or the other, for instance `git push origin gpu_cholesky:gpu_cholesky_sync` if you want to push the local branch `gpu_cholesky` to remote branch `gpu_cholesky_sync`. When you want to update `gpu_cholesky` being tracked by the PR, then you can push to it.

matt-graham commented:
@lamblin Thanks for the advice, I've set up two separate branches as you suggest.

nouiz commented:
For your other question, yes, you understood correctly my comment.

thanks


On Tue, Feb 16, 2016 at 1:05 PM, Matt Graham <notifications@github.com>
wrote:

> @lamblin <https://github.com/lamblin> Thanks for the advice, I've set up
> two separate branches as you suggest.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/4006#issuecomment-184803848>.
>

matt-graham commented:
@nouiz Now updated the `local_optimizer` decorator.
